services/common/auth_tools.py

```diff
@@ -0,0 +1,398 @@
+    if not organizations:
+        return False
+    allowed_orgs_str = ", ".join(f"'{org}'" for org in organizations)
+    query = (
+        "SELECT u.email as email "
+        "FROM public.users u "
+        "JOIN public.user_organization AS user_org ON user_org.user_id = intersection.user_id "
+        "JOIN public.organizations AS org ON org.organization_id = user_org.organization_id "
+        f"WHERE org.name = ANY (ARRAY[{allowed_orgs_str}])"
+        f"AND u.email = '{user_email}'"
+    )
+
+    logging.debug(f'Executing query: "{query};"')
+    data = pgquery.query_db(query)
+
+    return data[0]["email"] == user_email if data else False
+
+
+def get_index_or_default(
+    lst: list[ORG_ROLE_LITERAL], item: ORG_ROLE_LITERAL, default=-1
+) -> int:
+    try:
+        return lst.index(item)
+    except ValueError:
+        return default
+
+
+def check_role_above(
+    user_role: ORG_ROLE_LITERAL, required_role: ORG_ROLE_LITERAL
+) -> bool:
+    roles = [
+        ORG_ROLE_LITERAL.USER,
+        ORG_ROLE_LITERAL.OPERATOR,
+        ORG_ROLE_LITERAL.ADMIN,
+    ]
+    return get_index_or_default(roles, user_role) >= get_index_or_default(
+        roles, required_role
+    )
+
+
+def get_qualified_org_list(
+    user: EnvironWithOrg, required_role: ORG_ROLE_LITERAL, include_super_user=False
+) -> list[str]:
+    if include_super_user and user.user_info.super_user:
+        return pgquery.query_and_return_list(
+            "SELECT name FROM public.organizations ORDER BY name ASC"
+        )
+    allowed_orgs = []
+    for org_name, org_role in user.user_info.organizations.items():
+        if check_role_above(org_role, required_role):
+            allowed_orgs.append(org_name)
+    return allowed_orgs
+
+
+def protect_user_access(user_email: str, user: EnvironWithOrg):
+    if user_email != user.user_info.email:
+        qualified_orgs = get_qualified_org_list(
+            user, ORG_ROLE_LITERAL.ADMIN, include_super_user=False
+        )
+        if not user.user_info.super_user and not check_user_with_org(
+            user_email, qualified_orgs
+        ):
+            raise Forbidden(
+                f"403 Forbidden: User does not have access to modify notifications for user {user_email}"
+            )
+    return True
+
+
+class PermissionResult:
+    def __init__(
+        self,
+        allowed: bool,
+        qualified_orgs: list[str],
+        message: Optional[str],
+        user: EnvironWithOrg,
+    ):
+        self.allowed = allowed
+        self.qualified_orgs = qualified_orgs
+        self.message = message
+        self.user = user
+
+    def to_dict(self):
+        return self.__dict__
+
+
+class PermissionChecker(Protocol):
+    def check(
+        self,
+        user: EnvironWithOrg,
+        required_role: ORG_ROLE_LITERAL | None,
+        resource_type: Optional[RESOURCE_TYPE] = None,
+        resource_id: Optional[str] = None,
+    ): ...
+
+
+@dataclass
+class DefaultPermissionChecker:
```

Comment:
suggestion: @dataclass doesn't seem to be needed here as DefaultPermissionChecker doesn't have any fields